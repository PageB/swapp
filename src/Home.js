import React, { useState } from 'react';
import { ThemeProvider } from './contexts/ThemeContext';

import { NavigationToolbar } from './components';
import Screens from './screens';
import Login from './screens/Login/Login';

const Home = () => {
  const [token, setToken] = useState(localStorage.getItem('token'));
  const [theme, setTheme] = useState('DarkTheme');

  /**
   * Set token value after user sing in successfully.
   *
   * @method handleLogin
   * @param {String} token
   */
  const handleLogin = token => {
    setToken(token);
  };

  /**
   * Reset token value after user sing out successfully.
   *
   * @method handleLogout
   */
  const handleLogout = () => {
    setToken('');
    localStorage.removeItem('token');
  };

  /**
   * Change body class names based on theme selection
   *
   * @method themeChangeHandler
   */
  const themeChangeHandler = () => {
    if (theme === 'DarkTheme') {
      document.body.classList.add('LightTheme');
      document.body.classList.remove('DarkTheme');
      setTheme('LightTheme');
    } else {
      document.body.classList.add('DarkTheme');
      document.body.classList.remove('LightTheme');
      setTheme('DarkTheme');
    }
  };

  return (
    <ThemeProvider value={theme}>
      <NavigationToolbar
        themeChanged={themeChangeHandler}
        userAuthenticated={!!token}
        onLogout={handleLogout}
      />
      {token ? <Screens /> : <Login onLogin={handleLogin} />}
    </ThemeProvider>
  );
};

export default Home;
