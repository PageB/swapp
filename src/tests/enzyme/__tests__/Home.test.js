import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure, mount, render } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import { MemoryRouter } from 'react-router-dom';

// The component AND the query need to be exported
import Home from '../../../Home';
import gql from 'graphql-tag.macro';
const SIGN_IN = gql`
  mutation SignIn($email: String!, $password: String!) {
    signIn(email: $email, password: $password) {
      token
    }
  }
`;

const mocks = [
  {
    request: {
      query: SIGN_IN,
      variables: {
        email: 'martin@mail.com',
        password: 'martin'
      },
    },
    result: {
      data: {
        token: 'adfadfadsgdashahashah'
      },
    },
  },
];

configure({ adapter: new Adapter() });

describe('Home Component', () => {
  it('should match snapshot with Login component', () => {
    const wrapper = shallow(<Home />);

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Home Component', () => {
  it('should match snapshot with Screen component', () => {
    localStorage.setItem('token', '12312423423141241341')
    const wrapper = shallow(<Home />);

    expect(wrapper).toMatchSnapshot();
    localStorage.removeItem('token')
  });
});

describe('Home Component is rendered', () => {
  it('should remove token in local storage when user logout', () => {
    localStorage.setItem('token', '12312423423141241341')
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <MemoryRouter>
          <Home />
        </MemoryRouter>
      </MockedProvider>);

    wrapper.find('a').at(2).simulate('click');

    expect(localStorage.getItem('token')).toBe(null);
  });
});

// TODO: NEEDS TO BE FIXED
// describe('Home Component is rendered', () => {
//   it('should set token in local storage when user login', () => {
//     const preventDefault = jest.fn();
//     const wrapper = mount(
//       <MockedProvider mocks={mocks} removeTypename={true}>
//         <MemoryRouter>
//           <Home />
//         </MemoryRouter>
//       </MockedProvider>);

//     wrapper.find('form').simulate('submit', { preventDefault });

//     expect(localStorage.getItem('token')).toBe('adfadfadsgdashahashah');
//   });
// });

describe('Home Component is rendered', () => {
  it('should change theme when Navigation Icon is clicked', () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <MemoryRouter>
          <Home />
        </MemoryRouter>
      </MockedProvider>);

    expect(wrapper.find('.Toolbar').hasClass('Toolbar DarkTheme')).toBe(true);

    wrapper.find('.Toolbar__Logo').simulate('click');

    expect(wrapper.find('.Toolbar').hasClass('Toolbar LightTheme')).toBe(true);

    wrapper.find('.Toolbar__Logo').simulate('click');

    expect(wrapper.find('.Toolbar').hasClass('Toolbar DarkTheme')).toBe(true);
  });
});

