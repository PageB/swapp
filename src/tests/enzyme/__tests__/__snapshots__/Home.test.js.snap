// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should match snapshot with Login component 1`] = `
<ContextProvider
  value="DarkTheme"
>
  <NavigationToolbar
    onLogout={[Function]}
    theme="DarkTheme"
    themeChanged={[Function]}
    userAuthenticated={false}
  />
  <Login
    onLogin={[Function]}
  />
</ContextProvider>
`;

exports[`Home Component should match snapshot with Screen component 1`] = `
<ContextProvider
  value="DarkTheme"
>
  <NavigationToolbar
    onLogout={[Function]}
    theme="DarkTheme"
    themeChanged={[Function]}
    userAuthenticated={true}
  />
  <Screens />
</ContextProvider>
`;
